File: dasan_cli/status.py
"""Status & network overview: device info, WAN, LAN, DHCP clients."""
import typer
from typing_extensions import Annotated

from . import core

app = typer.Typer(help="Device status, WAN/LAN, and connected clients")


@app.command()
def info(ctx: typer.Context):
    """Device info, uptime, CPU/mem/temp, PON optical stats."""
    client = ctx.obj
    info = client.get("DeviceInfo")
    pon = client.get("PonPortStatus")
    core.console.print(f"[bold]Model:[/bold]        {info['manufacturer']} {info['modelName']}")
    core.console.print(f"[bold]Serial:[/bold]       {info['serialNumber']}")
    core.console.print(f"[bold]HW/SW ver:[/bold]    {info['hardwareVersion']} / {info['softwareVersion']}")
    core.console.print(f"[bold]Uptime:[/bold]       {info['UpTime']}s ({info['UpTime']//3600}h{(info['UpTime']%3600)//60}m)")
    core.console.print(f"[bold]CPU / Mem:[/bold]    {info['CPU_Load']}% / {info['MemoryUsage']}%")
    core.console.print(f"[bold]Temperature:[/bold]  {info['Temperature']} C")
    core.console.print(f"[bold]MAC:[/bold]          {info['MACAddress']}")
    core.console.print()
    core.console.print(f"[bold]PON:[/bold]          {pon['ponMode']} link {pon['ponLinkState']} "
                        f"(uptime {pon['ponLinkUptime']}s), OLT {pon['oltType']}")
    core.console.print(f"[bold]Optical:[/bold]      Rx {pon['ponRxPower']} dBm / Tx {pon['ponTxPower']} dBm, "
                        f"temp {pon['ponTemp']} C, FEC {pon['fecStatus']}")


@app.command()
def wan(ctx: typer.Context):
    """WAN connection summary (internet/voice VLANs, IPs, status)."""
    client = ctx.obj
    conns = client.get("WANIPConnection")
    rows = []
    for c in conns:
        if not c.get("enable") and c.get("connectionType") == "Unconfigured":
            continue
        rows.append([
            c["iid"], c["ServiceList"], c["connectionType"], c["connectionStatus"],
            c["externalIPAddress"] or "-", c["defaultGateway"] or "-", c["VLANId"],
        ])
    core.render_table(rows, ["iid", "service", "type", "status", "ip", "gateway", "vlan"])


@app.command()
def lan(ctx: typer.Context):
    """LAN port link status."""
    client = ctx.obj
    ports = client.get("LANPortStatus")
    rows = [[p["iid"], p["Admin"], p["Status"], p["Mode"]] for p in ports]
    core.render_table(rows, ["port", "admin", "status", "mode"])


@app.command()
def clients(ctx: typer.Context):
    """DHCP leases / connected devices."""
    client = ctx.obj
    leases = client.get("DhcpLease")
    rows = [[l["ClientName"], l["IP"], l["MAC"], l["ExpireTime"]] for l in leases]
    core.render_table(rows, ["name", "ip", "mac", "lease-expires"])

File: internal/cli/status.go
package cli

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/anshuman852/dasan/internal/client"
)

// NewStatusCmd creates the `dasan status` command tree.
func NewStatusCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "status",
		Short: "Device status, WAN/LAN, and connected clients",
	}

	cmd.AddCommand(
		newStatusInfoCmd(),
		newStatusWanCmd(),
		newStatusLanCmd(),
		newStatusClientsCmd(),
	)

	return cmd
}

func newStatusInfoCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "info",
		Short: "Device info, uptime, CPU/mem/temp, PON optical stats",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			info, err := cl.Get("DeviceInfo", "")
			if err != nil {
				return err
			}
			pon, err := cl.Get("PonPortStatus", "")
			if err != nil {
				return err
			}

			obj, _ := client.ToObj(info)
			ponObj, _ := client.ToObj(pon)

			fmt.Printf("Model:         %s %s\n", client.GetStr(obj, "manufacturer"), client.GetStr(obj, "modelName"))
			fmt.Printf("Serial:        %s\n", client.GetStr(obj, "serialNumber"))
			fmt.Printf("HW/SW ver:     %s / %s\n", client.GetStr(obj, "hardwareVersion"), client.GetStr(obj, "softwareVersion"))
			uptime := client.GetFloat(obj, "UpTime")
			fmt.Printf("Uptime:        %.0fs (%dh%dm)\n", uptime, int(uptime)/3600, int(uptime)%3600/60)
			fmt.Printf("CPU / Mem:     %.0f%% / %.0f%%\n", client.GetFloat(obj, "CPU_Load"), client.GetFloat(obj, "MemoryUsage"))
			fmt.Printf("Temperature:   %.0f C\n", client.GetFloat(obj, "Temperature"))
			fmt.Printf("MAC:           %s\n", client.GetStr(obj, "MACAddress"))
			fmt.Println()
			fmt.Printf("PON:           %s link %s (uptime %.0fs), OLT %s\n",
				client.GetStr(ponObj, "ponMode"), client.GetStr(ponObj, "ponLinkState"),
				client.GetFloat(ponObj, "ponLinkUptime"), client.GetStr(ponObj, "oltType"))
			fmt.Printf("Optical:       Rx %s dBm / Tx %s dBm, temp %s C, FEC %s\n",
				client.GetStr(ponObj, "ponRxPower"), client.GetStr(ponObj, "ponTxPower"),
				client.GetStr(ponObj, "ponTemp"), client.GetStr(ponObj, "fecStatus"))
			return nil
		},
	}
}

func newStatusWanCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "wan",
		Short: "WAN connection summary (internet/voice VLANs, IPs, status)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("WANIPConnection", "")
			if err != nil {
				return err
			}
			arr, err := client.ToArr(data)
			if err != nil {
				return err
			}
			rows := [][]string{}
			for _, item := range arr {
				c, ok := item.(map[string]any)
				if !ok {
					continue
				}
				if !client.GetBool(c, "enable") && client.GetStr(c, "connectionType") == "Unconfigured" {
					continue
				}
				ip := client.GetStr(c, "externalIPAddress")
				if ip == "" {
					ip = "-"
				}
				gw := client.GetStr(c, "defaultGateway")
				if gw == "" {
					gw = "-"
				}
				rows = append(rows, []string{
					Ftoa(client.GetFloat(c, "iid")),
					client.GetStr(c, "ServiceList"),
					client.GetStr(c, "connectionType"),
					client.GetStr(c, "connectionStatus"),
					ip, gw,
					Ftoa(client.GetFloat(c, "VLANId")),
				})
			}
			PrintTable([]string{"iid", "service", "type", "status", "ip", "gateway", "vlan"}, rows)
			return nil
		},
	}
}

func newStatusLanCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "lan",
		Short: "LAN port link status",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("LANPortStatus", "")
			if err != nil {
				return err
			}
			arr, err := client.ToArr(data)
			if err != nil {
				return err
			}
			rows := [][]string{}
			for _, item := range arr {
				p, ok := item.(map[string]any)
				if !ok {
					continue
				}
				rows = append(rows, []string{
					Ftoa(client.GetFloat(p, "iid")),
					client.GetStr(p, "Admin"),
					client.GetStr(p, "Status"),
					client.GetStr(p, "Mode"),
				})
			}
			PrintTable([]string{"port", "admin", "status", "mode"}, rows)
			return nil
		},
	}
}

func newStatusClientsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "clients",
		Short: "DHCP leases / connected devices",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("DhcpLease", "")
			if err != nil {
				return err
			}
			arr, err := client.ToArr(data)
			if err != nil {
				return err
			}
			rows := [][]string{}
			for _, item := range arr {
				l, ok := item.(map[string]any)
				if !ok {
					continue
				}
				rows = append(rows, []string{
					client.GetStr(l, "ClientName"),
					client.GetStr(l, "IP"),
					client.GetStr(l, "MAC"),
					client.GetStr(l, "ExpireTime"),
				})
			}
			PrintTable([]string{"name", "ip", "mac", "lease-expires"}, rows)
			return nil
		},
	}
}
